File: webforj-components/webforj-html-elements/src/main/java/com/webforj/component/html/elements/Fieldset.java
package com.webforj.component.html.elements;

import com.webforj.component.Component;
import com.webforj.component.element.annotation.NodeName;
import com.webforj.component.html.HtmlComponentContainer;

/**
 * Component representing a {@code fieldset} element.
 *
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset">Html fieldset
 *      Tag</a>
 *
 * @author Hyyan Abo Fakher
 * @since 23.06
 */
// We're purposefully ignoring the deep inheritance warning here because we've designed our class
// hierarchy to meet the unique requirements of our UI framework. This design closely aligns with
// our framework's specific goals and emphasizes the need for caution when considering any changes.
//
// Any changes to the inheritance structure should be thoughtfully evaluated in the context of our
// framework's needs. The current structure is essential for meeting those needs.
@SuppressWarnings("squid:S110")
@NodeName("fieldset")
public class Fieldset extends HtmlComponentContainer<Fieldset> {

  /**
   * Creates a new empty fieldset.
   */
  public Fieldset() {
    super();
  }

  /**
   * Creates a new fieldset with the given text as {@link Legend}.
   *
   * @param text the legend's text
   */
  public Fieldset(String text) {
    this();
    add(new Legend(text));
  }

  /**
   * Creates a new fieldset with the given child components.
   *
   * @param components the child components
   */
  public Fieldset(Component... components) {
    super(components);
  }
}

File: webforj-kotlin/src/main/kotlin/com/webforj/kotlin/dsl/component/html/elements/FieldSet.kt
package com.webforj.kotlin.dsl.component.html.elements

import com.webforj.component.html.elements.Fieldset
import com.webforj.component.html.elements.Legend
import com.webforj.concern.HasComponents
import com.webforj.kotlin.dsl.WebforjDsl
import com.webforj.kotlin.dsl.init

/**
 * Creates a `FieldSet` with an optional [text].
 * ```
 * ... {
 *   Fieldset() // Empty fieldset element
 *   Fieldset("text") // fieldset element with legend element
 * }
 * ```
 *
 * @param text The text to add to the `Fieldset` as [Legend].
 * @param block The initialization steps for the `Fieldset`.
 * @return The configured `Fieldset` instance.
 * @see Fieldset
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset">Html fieldset
 *       Tag</a>
 */
fun @WebforjDsl HasComponents.fieldset(text: String? = null, block: @WebforjDsl Fieldset.() -> Unit = {}): Fieldset {
    val fieldSet = text?.let { Fieldset(text) } ?: Fieldset()
    return init(fieldSet, block)
}

/**
 * Creates a `Legend` with an optional [text].
 * ```
 * fieldset {
 *   legend() // Empty legend element
 *   legend("text") // legend element with text
 * }
 * ```
 *
 * @param text The text to add to the `Legend`.
 * @param block The initialization steps for the `Legend`.
 * @return The configured `Legend` instance.
 * @see Legend
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/legend">HTML legend
 *      Tag</a>
 */
fun @WebforjDsl Fieldset.legend(text: String? = null, block: @WebforjDsl Legend.() -> Unit = {}): Legend {
    val legend = text?.let { Legend(text) } ?: Legend()
    return init(legend, block)
}
